### ISSUE TITLE
 after 8 epochs the bot is still saying "I don't know" despite being on latest commit.

### ISSUE THREAD

**Issue by huwprosser** [fawazsammani/chatbot-transformer#2]
Have been training for 8 epochs. System only recommend "I dont know". I imagine this is linked to the previous Issue request but still seems to be a bug in master branch latest commit.

**Comment by fawazsammani** [comment]
Alright, I will train it and see if i face the same problem. If yes, I will fix it and update you.
Thanks for letting me know about that bug

**Comment by huwprosser** [comment]
Thank you very much. Really appreciate the reply.

**Comment by huwprosser** [comment]
We're there any further developments?

**Comment by fawazsammani** [comment]
Dear @huwissp,
I'm currently busy with a conference submission. After I submit, then I will be free to check the repo again. In the meantime, you can go through it and check for errors

[UPDATE] Please see my latest comment below [here](https://github.com/fawazsammani/chatbot-transformer/issues/2#issuecomment-1943484438)

**Comment by j-v-n** [comment]
@huwissp  I was facing the same issue. I ran the universal transformer chatbot jupyter notebook and the chatbot works fine. but if i run the transformer chatbot jupyter notebook, it replies 'i dont know' to almost every question. I haven't yet checked to see what the differences are between the two notebooks. Hope this helps

**Comment by Sri-user** [comment]
This is not a question, but does anyone has done a keras implementation of the same?

**Comment by yousrajnahi** [comment]
@j-v-n same problem, did you run the universal transformer chatbot using google colab ? and how much epochs?

**Comment by huakeda1** [comment]
I come across the same problem that it always gives a reply about i don't know as for almost every question, any one find the problem?

**Comment by harrywillss** [comment]
Having the same issue after 10 epochs. Anyone have a fix?

**Comment by asarra** [comment]
*push*

**Comment by fitzjalen** [comment]
Same problem. Even if I use different architecture on the same preprocessing.

**Comment by guochen-code** [comment]
@fawazsammani do you have a chance to figure out whether it is a bug at all?

Only get 'i dont know' repy for both base transformer and universal transformer, unless reduce the number of layers to 1, then it can reply back something different, but overall the results are not satisfied. 

Is transformer really powerful in this application? I found using seq2seq generates much better and meaningful replies.

**Comment by MessiNN** [comment]
While, I was checking the code I found that the chat bot says "I don't know" only when I use the "chat.py" file but it works when I use the "transformer chatbot.ipynb" file. I don't know exactly what is causing the chat bot to say "i don't know" but if you want a quick fix; Copy paste some parts of the "transformer chatbot.ipynb" into the chat file. I will continue checking for some time.

**Comment by Pablito2008** [comment]
Hello, the model is still saying 'i dont know', does anyone have a solution please ?
thank you

**Comment by newmosesx** [comment]
Just keep on training it. Check the number of losses. Modify the layers. Basically experiment with it.
There is no quick solution when it comes to this.

**Comment by MessiNN** [comment]
Try reducing loss until  you reach between 3.5 to 2.5. The more you specialise your chat bot by reducing your loss might allow you to have a much more accurate response.  Obviously, if you specialised it too much might become a problem as it will not longer think outside the box but only on what you trained it on.

**Comment by MessiNN** [comment]
I made a few modifications to the code but for the entire part it is the same. Once I feel confident in them I might share them later, I won't give you my models since the files will contain everything needed for it. Just keep training and increase training data repeatedly and it should work.

**Comment by MessiNN** [comment]
@fawazsammani  If anything wait for the owner to have some free time, I guess.

**Comment by MessiNN** [comment]
Also don't just ask for help. Go into the world and learn.
Recomendation:
https://youtube.com/playlist?list=PLTl9hO2Oobd97qfWC40gOSU8C0iu0m2l4&si=bRuPHpT8-eTpqv-Q

**Comment by fawazsammani** [comment]
Dear all,
I am very sorry for the late reply. The problem occurs mainly because of overfitting. Transformer models have no inductive bias as in LSTMs. They are more generic since they are only composed of MLPs and thus require much more training data to work. This is a very well known issue, and the power of Transformer really comes from pre-training and fine-tuning. That is why LSTMs work much better on this small dataset. In any case, to solve this issue, you need to avoid overfitting: Reduce the dimensionality of the layers (dim 128 or 256 should be OK) and/or reduce the number of layers to a maximum of 3. Otherwise, if you want to leave it with the current settings, you need to train it on more data.

Ofcourse you would not have to deal with overfitting if you first pretrain the model (or load any pretrained LM, there are plenty of them on hugging face - for example Distilled-GPT2 has the same setting as the one we are doing in this tutorial) and then fine-tune it on this small dataset. Because Transfer learning is itself one of the best ways to combat overfitting, and is the main reason why Transformers work much better than any other model. This will lead to the best results. 

You can also use any pretrained Transformer Encoder (e.g., Sentence-BERT models) and freeze it, and only train/fine-tune the transformer decoder. This will lead to even better results and save you more memory.

Finally note that ChatGPT is finetuned on conversational data of instructions (commonly referred to as Instruction datasets). The dataset we use here is considered as conversational data, but not in the form of Instructions. Nevertheless, the performance of ChatGPT rises after the finetuning, and therefore, taking a pretrained LM (which are abudantly available on hugging face - of different sizes) and finetuning it on this conversational dataset we use here in this tutorial, will cause your chatbot to shine.

Regards
